flash/core/optimizers/lamb.py

```diff
@@ -0,0 +1,162 @@
+        """
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        for group in self.param_groups:
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+                grad = p.grad
+                if grad.is_sparse:
+                    raise RuntimeError("LAMB does not support sparse gradients")
+                amsgrad = group["amsgrad"]
+                exclude_from_layer_adaptation = group["exclude_from_layer_adaptation"]
+
+                state = self.state[p]
+
+                # State initialization
+                if len(state) == 0:
+                    state["step"] = 0
+                    # Exponential moving average of gradient values
+                    state["exp_avg"] = torch.zeros_like(p, memory_format=torch.preserve_format)
+                    # Exponential moving average of squared gradient values
+                    state["exp_avg_sq"] = torch.zeros_like(p, memory_format=torch.preserve_format)
+                    if amsgrad:
+                        # Maintains max of all exp. moving avg. of sq. grad. values
+                        state["max_exp_avg_sq"] = torch.zeros_like(p, memory_format=torch.preserve_format)
+
+                exp_avg, exp_avg_sq = state["exp_avg"], state["exp_avg_sq"]
+                if amsgrad:
+                    max_exp_avg_sq = state["max_exp_avg_sq"]
+                beta1, beta2 = group["betas"]
+
+                state["step"] += 1
+                bias_correction1 = 1 - beta1 ** state["step"]
+                bias_correction2 = 1 - beta2 ** state["step"]
+
+                # Decay the first and second moment running average coefficient
+                exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+                exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)
+                if amsgrad:
+                    # Maintains the maximum of all 2nd moment running avg. till now
+                    torch.max(max_exp_avg_sq, exp_avg_sq, out=max_exp_avg_sq)
+                    # Use the max. for normalizing running avg. of gradient
+                    denom = (max_exp_avg_sq.sqrt() / math.sqrt(bias_correction2)).add_(group["eps"])
+                else:
+                    denom = (exp_avg_sq.sqrt() / math.sqrt(bias_correction2)).add_(group["eps"])
+
+                numerator = exp_avg / bias_correction1
+                update = numerator / denom
+
+                if group["weight_decay"] != 0:
+                    update = update.add(p.data, alpha=group["weight_decay"])
+
+                trust_ratio = 1.0
+                if not exclude_from_layer_adaptation:
+                    w_norm = torch.norm(p.data)
+                    g_norm = torch.norm(update)
+
+                    if w_norm > 0 and g_norm > 0:
+                        trust_ratio = w_norm / g_norm
+
+                p.add_(update, alpha=-group["lr"] * trust_ratio)
```

Comment:
can we wrap this to some internal protected method?

Reply:
right now it works, I don't wanna change the code for an optimizer implementation which works, takes too much time to delve into it and make sure things do not get messed up.